File: engine/glyphs.py
"""Glyph access layer for Compresso Parametric Studio.

Matrices live in ``glyphs_data.py`` (Compresso import) with hand overrides
from ``glyph_overrides.py`` (TYPE TOOL corrections).
"""

from __future__ import annotations

from typing import Iterable

from engine.glyph_overrides import GLYPH_OVERRIDES, WIDTH_OVERRIDES
from engine.glyphs_data import GLYPHS as _RAW_GLYPHS
from engine.glyphs_data import GLYPH_WIDTHS as _RAW_WIDTHS

ROWS_TOTAL = 28
ACCENT_TOP, ACCENT_BOTTOM = 0, 3
BODY_TOP, BODY_BOTTOM = 4, 23
DESC_TOP, DESC_BOTTOM = 24, 27
CAP_HEIGHT = BODY_BOTTOM - BODY_TOP + 1  # 20
BASELINE = BODY_BOTTOM  # 23
SPACE_WIDTH_COLS = 5

GLYPH_CHARS: list[str] = (
    [chr(c) for c in range(ord("A"), ord("Z") + 1)]
    + list("АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ")
    + list("0123456789")
    + list(".,:;!?/+-=")
)

GLYPH_WIDTHS: dict[str, int] = {**_RAW_WIDTHS, **WIDTH_OVERRIDES}

GLYPHS: dict[str, list[tuple[int, int]]] = {
    ch: list(pts) for ch, pts in _RAW_GLYPHS.items()
}
GLYPHS.update({ch: list(pts) for ch, pts in GLYPH_OVERRIDES.items()})


def resolve_glyph_key(ch: str) -> str | None:
    """Map input character to a ``GLYPHS`` key, or ``None`` for whitespace."""
    if ch in " \t":
        return " "
    if ch in "\n\r\v\f":
        return None
    if ch in "ёЁ":
        return "Ё"
    upper = ch.upper()
    if upper in GLYPHS:
        return upper
    if ch in GLYPHS:
        return ch
    return ""


def normalize_text(text: str) -> list[str]:
    """Normalize user text to All-Caps glyph keys; unknowns become empty tokens."""
    out: list[str] = []
    for ch in text:
        key = resolve_glyph_key(ch)
        if key is None:
            continue
        out.append(key)
    return out


def glyph_width(ch: str) -> int:
    """Advance width in grid columns for ``ch``."""
    if ch == " ":
        return SPACE_WIDTH_COLS
    if ch == "":
        return SPACE_WIDTH_COLS
    if ch in GLYPH_WIDTHS:
        return max(1, GLYPH_WIDTHS[ch])
    pts = GLYPHS.get(ch, [])
    if not pts:
        return SPACE_WIDTH_COLS
    return max(c for c, _ in pts) + 1


def scaled_width(ch: str, col_scale: int) -> int:
    """Column width after matrix density scaling."""
    return glyph_width(ch) * max(1, col_scale)


def scale_glyph_density(
    coords: Iterable[tuple[int, int]],
    src_cols: int,
    col_scale: int,
    row_scale: int,
) -> list[tuple[int, int]]:
    """Upsample glyph modules for matrix density sliders."""
    if col_scale <= 1 and row_scale <= 1:
        return list(coords)
    out: set[tuple[int, int]] = set()
    for c, r in coords:
        for dr in range(row_scale):
            for dc in range(col_scale):
                out.add((c * col_scale + dc, r * row_scale + dr))
    if row_scale > 1:
        max_r = ROWS_TOTAL * row_scale - 1
        squashed: set[tuple[int, int]] = set()
        for c, r in out:
            squashed.add((c, int(r * (ROWS_TOTAL - 1) / max_r)))
        out = squashed
    return sorted(out, key=lambda t: (t[1], t[0]))


def get_glyph(
    ch: str,
    col_scale: int = 1,
    row_scale: int = 1,
) -> list[tuple[int, int]]:
    """Return module coordinates for ``ch``; unknown/space yields empty list."""
    key = resolve_glyph_key(ch) if len(ch) == 1 else ch
    if key is None or key == " " or key == "":
        return []
    if key not in GLYPHS:
        return []
    base = GLYPHS[key]
    if not base or (col_scale <= 1 and row_scale <= 1):
        return list(base)
    return scale_glyph_density(base, glyph_width(key), col_scale, row_scale)

File: src/engine/glyphs.ts
/**
 * All-Caps glyph access layer.
 *
 * The grid is 28 rows tall:
 *   rows  0..3   accents (Ё, Й diacritics)
 *   rows  4..23  cap-height body (20 rows)
 *   row     23   baseline — every glyph in a line sits on it
 *   rows 24..27  descenders (Ц, Щ tails, Д legs)
 */

import {
  BASELINE,
  GLYPHS,
  GLYPH_CHARS,
  GLYPH_WIDTHS,
  ROWS_TOTAL,
  SPACE_WIDTH_COLS,
} from '../data/glyphsData';
import type { GlyphMatrix, GridCoord } from '../types/fontTypes';

export {
  ACCENT_BOTTOM,
  ACCENT_TOP,
  BASELINE,
  BODY_BOTTOM,
  BODY_TOP,
  CAP_HEIGHT,
  DESC_BOTTOM,
  DESC_TOP,
  GLYPHS,
  GLYPH_CHARS,
  GLYPH_WIDTHS,
  ROWS_TOTAL,
  SPACE_WIDTH_COLS,
} from '../data/glyphsData';

const EMPTY: GlyphMatrix = [];

/** Space or an unsupported character — laid out as an advance with no ink. */
export function isBlank(ch: string): boolean {
  return ch === ' ' || ch === '';
}

/**
 * Map an input character onto a glyph key: whitespace folds to a space,
 * line breaks are dropped, `ё` folds to `Ё`, everything else upper-cases.
 * Returns `null` for characters that should be skipped entirely.
 */
export function resolveGlyphKey(ch: string): string | null {
  if (ch === ' ' || ch === '\t') {
    return ' ';
  }
  if (ch === '\n' || ch === '\r' || ch === '\v' || ch === '\f') {
    return null;
  }
  const folded = ch === 'ё' || ch === 'Ё' ? 'Ё' : ch.toUpperCase();
  return folded in GLYPHS ? folded : '';
}

/** All-Caps normalization: one entry per rendered character. */
export function normalizeText(text: string): string[] {
  const out: string[] = [];
  for (const ch of text) {
    const key = resolveGlyphKey(ch);
    if (key !== null) {
      out.push(key);
    }
  }
  return out;
}

/** Advance width of a glyph in grid columns. */
export function glyphWidth(ch: string): number {
  if (isBlank(ch)) {
    return SPACE_WIDTH_COLS;
  }
  const known = GLYPH_WIDTHS[ch];
  if (known !== undefined) {
    return Math.max(1, known);
  }
  const coords = GLYPHS[ch];
  if (!coords || coords.length === 0) {
    return SPACE_WIDTH_COLS;
  }
  let maxCol = 0;
  for (const [col] of coords) {
    if (col > maxCol) {
      maxCol = col;
    }
  }
  return maxCol + 1;
}

export function scaledWidth(ch: string, colScale: number): number {
  return glyphWidth(ch) * Math.max(1, colScale);
}

/**
 * Multiply module density: every cell becomes a `colScale × rowScale` block.
 * Rows are squashed back into the 28-row grid so the baseline stays put.
 */
function scaleGlyphDensity(
  coords: GlyphMatrix,
  colScale: number,
  rowScale: number,
): GlyphMatrix {
  const expanded = new Map<number, GridCoord>();
  for (const [col, row] of coords) {
    for (let dr = 0; dr < rowScale; dr += 1) {
      for (let dc = 0; dc < colScale; dc += 1) {
        const c = col * colScale + dc;
        const r = row * rowScale + dr;
        expanded.set(c * 4096 + r, [c, r]);
      }
    }
  }

  let list = [...expanded.values()];

  if (rowScale > 1) {
    const maxRow = ROWS_TOTAL * rowScale - 1;
    const squashed = new Map<number, GridCoord>();
    for (const [col, row] of list) {
      const r = Math.round((row * (ROWS_TOTAL - 1)) / maxRow);
      squashed.set(col * 4096 + r, [col, r]);
    }
    list = [...squashed.values()];
  }

  list.sort((a, b) => a[1] - b[1] || a[0] - b[0]);
  return list;
}

const densityCache = new Map<string, GlyphMatrix>();

/** Module coordinates for one character at the given matrix multipliers. */
export function getGlyph(ch: string, colScale: number, rowScale: number): GlyphMatrix {
  const base = GLYPHS[ch];
  if (!base || base.length === 0) {
    return EMPTY;
  }
  if (colScale <= 1 && rowScale <= 1) {
    return base;
  }
  const key = `${ch}|${colScale}|${rowScale}`;
  const cached = densityCache.get(key);
  if (cached) {
    return cached;
  }
  const scaled = scaleGlyphDensity(base, colScale, rowScale);
  densityCache.set(key, scaled);
  return scaled;
}

/** Characters offered in the glyph inspector, in alphabet order. */
export const INSPECTOR_CHARS: readonly string[] = GLYPH_CHARS;

/** Row index of the baseline — exported for guide drawing. */
export const BASELINE_ROW = BASELINE;
